### ISSUE TITLE
Demo: live pix2pix (edges2cats)

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#79]
_From @nsthorat on October 19, 2017 15:34_

pix2pix demo was quite fun, and had a nice demo page:
https://affinelayer.com/pixsrv/

However, it is often down, and requires a TensorFlow backend, and if we port it to deeplearn.js you could see your output evolve as you draw, instead of after a round trip to a server.

The caveat here is that the model may be huge. The model itself doesn't take too long to train (on the order of a few hours with a GPU), but we can trade off model performance for number of parameters.

There are hyperparameters that are exposed in the TensorFlow implementation that allow you to change hyperparameters which may help make it shippable to the web, which you can see here:
https://github.com/affinelayer/pix2pix-tensorflow/blob/master/pix2pix.py#L16

_Copied from original issue: tensorflow/tfjs-core#239_

**Comment by nsthorat** [comment]
_From @yining1023 on December 28, 2017 22:37_

edges2cats is very cool! I'd like to port it to deeplearn.js.

I tried [pix2pix-tensorflow](https://github.com/affinelayer/pix2pix-tensorflow) and trained the facades dataset(The pre-trained model from pix2pix-tensorflow doesn't work for me). And I got an `index.html` that shows input/output/target image sets. The result looks good.

But the cat dataset is not available. So I think we can prepare our own dataset, and it doesn't need to be about cats. What do you think? Any suggestions?

<img src='https://user-images.githubusercontent.com/8662372/34424615-f4defc9e-ebf2-11e7-9432-a65337d6c938.jpeg' width='300'/>

**Comment by nsthorat** [comment]
_From @christopherhesse on December 29, 2017 2:34_

By a hilarious coincidence I am currently porting the afflinelayer demo to use deeplearnjs instead of a remote server, @yining1023 you can check out the code when it is done (probably tomorrow) if you want to make a interactive demo or just a better one.  It would be pretty cool if there was a version nice enough to put on this page: https://deeplearnjs.org/index.html#demos provided the deeplearnjs guys are cool with it.

@nsthorat the performance seems a bit lower than I expected.  Tensorflow inference (CPU only) is 300ms per image, looks like deeplearnjs is 5000ms for the first image and 2500ms thereafter in Chrome on Mac OS X.  Is there some trick to make that faster?  It mentions a WebGL context so I assume it's using GPU.  Also, the entire page becomes unresponsive when it's running the model.

I am not currently using graphs or sessions, which could be the reason, is that likely to make things faster?

If I use `const math = dl.ENV.getBackend("cpu")`, it just hangs on the first operation I do.

**Comment by nsthorat** [comment]
_From @yining1023 on December 29, 2017 4:6_

@christopherhesse Sounds great! I'd love to make an interactive demo.

**Comment by nsthorat** [comment]
_From @christopherhesse on December 29, 2017 4:10_

Awesome.

Also @nsthorat is there a method to tell if the current browser is supported?  Right now the demo just breaks mysteriously on safari and internet explorer, I would prefer to pop up some sort of error message, but the browser support is poorly documented and it seems that deeplearn.js should tell me if it works with the current browser rather than me doing some ad-hoc thing.

**Comment by nsthorat** [comment]
_From @christopherhesse on December 29, 2017 4:41_

@yining1023 Here's the new local demo mode docs: https://github.com/affinelayer/pix2pix-tensorflow/tree/master/server

**Comment by nsthorat** [comment]
Hi @christopherhesse. Took a look at your demo and code, here are some thoughts on making it faster, in increasing order of difficulty:

1) Instead of constructing a Float32Array when using the edges image, use `Array3D.fromPixels` and give the the canvas element (try to avoid using ImageData). This will move render data from canvas => WebGLTexture on the GPU and entirely bypass the CPU. https://github.com/affinelayer/pix2pix-tensorflow/blob/master/server/static/index.html#L491
An example: https://github.com/PAIR-code/deeplearnjs/blob/master/demos/imagenet/imagenet.ts#L187

2) Draw directly to the canvas and bypass calling getValuesAsync and using ImageData by:
  - Constructing a math object with the your canvas render target, like in the following code block (the caveat is that you can only have one output canvas, bound (I notice that you have multiple output canvases on the page for each of the models. You can either move that output canvas around, or render to the output canvas and use ctx.drawImage):
```
const gl = gpgpu_util.createWebGLContext(outputCanvas);
const gpgpu = new GPGPUContext(gl);
const backend = new MathBackendWebGL(gpgpu);
const math = new NDArrayMath(backend, false);
ENV.setMath(math);
```

  - Construct a render shader when the page loads:
```
const renderShader = render_ndarray_gpu_util.getRenderRGBShader(
    this.gpgpu, inferenceResult.shape[1]);  // columns of output
```
  - Render the image like this (instead of using getValuesAsync and ImageData like you're doing now):
```
render_ndarray_gpu_util.renderToCanvas(gpgpu, renderShader,
    this.backend.getTexture(outputArray.id));
```
3) Retrain your model with ops that are optimized for edge devices. Caveat: this will change GAN training dynamics, you will have to weaken your discriminator.
  - Instead of "conv2d" use "separable_conv2d": https://www.tensorflow.org/api_docs/python/tf/nn/separable_conv2d
  - Instead of "conv2dTranspose" use "separable_conv2d" at each layer - the separable_conv2d op here should map an image to another image with the same shape. The use bilinear resizing to double the image size.

#3 will give you the biggest wins if you're willing to retrain it with separable convs instead of dense convs. I highly recommend doing this as it will not only greatly speed up your inference time (we've seen up to 7x speed improvements in convolutional models), but it will vasty decrease the size of the .pict file (which is ~50MB in your demo right now which may be a little too large to ship on the web).

As for device support, Safari *should* work, though it will be much slower. How is it breaking? We should support all WebGL 1.0 devices, but Safari we have to use float32 => 4 byte integer packing so the format of the upload / download can throw things off.

**Comment by nsthorat** [comment]
_From @dsmilkov on December 29, 2017 14:47_

> 2. Draw directly to the canvas and bypass calling getValuesAsync and using ImageData by

I would lower the priority on point 2) above due to its complexity and not giving much potential win. There could be also problems with flickering if you don't wait for the gpu to fully finish before processing the next frame.

> 3. Retrain your model with ops that are optimized for edge devices. Caveat: this will change GAN training dynamics, you will have to weaken your discriminator.

+1. This will definitely give the biggest win and worth pursing first.

**Comment by nsthorat** [comment]
_From @christopherhesse on December 29, 2017 19:7_

Thanks for the tips!

1 + 2: If I just remove my model (`var output_rgb = input_rgb`) then it takes about 100ms, so I don't think the bottleneck is messing with ImageData or getValuesAsync.  Is that likely to be the case?

3:  This sounds promising,  I will look into it.  50MB was around the maximum size I wanted to ship, but is not actually that bad compared to the size of videos.  Do you have any idea why my model is slower in the browser vs CPU/GPU?  I looked into using a graph in case creating that was slow, but the graph object had different arguments and was missing functions compared to the NDArrayMath object.

browser dl.js: 2500ms
CPU TF: 300ms
GPU TF: 50ms

Device support: My bad, Safari does work.  It has a weird pause, but it's likely to be my code.  I tested Internet Explorer and Edge, which may have failed for reasons unrelated to deeplearn.js, I didn't really investigate much.  So IE and Edge should work fine right?

**Comment by nsthorat** [comment]
It's pretty surprising that it's that much of a difference, but TF has shared memory and optimized kernels (eigen C++ is much much more mature than we are).

GPU TF, with CUDA, is insanely optimized / close to the metal. Unfortunately OpenGL is strictly slower because it's not designed to do linear algebra. However, in theory it should be on par with CPU TF. My guess is that one of the major bottlenecks is the dense transposed convolutions are super optimized in CPU TF (we haven't payed a ton of attention to conv2dTranspose because separable convolutions + bilinear resizing are way faster).

If you get reasonable performance (in terms of how well it reproduces the original image) on a model after trying #3, let me know and I can drill down deeper and see what inference performance bottlenecks are!

IE / Edge should work, but I haven't looked at them in a while (we haven't yet set up integration testing for them).

**Comment by nsthorat** [comment]
_From @christopherhesse on December 30, 2017 22:5_

I did a separable conv then bilinear filter and it didn't work.  Based on a distill.pub article, i tried nearest neighbor then separable conv and it worked, but the quality seemed lower, though the param count was reduced considerably.  It may require more tweaks to get better output.  I'm going to leave my current implementation as-is since I don't have time to try more tweaks right now.  https://github.com/affinelayer/pix2pix-tensorflow/commit/0f2174424356162920fc25067e63db2bcf19f4e7

Did you guys have any interest in an interactive demo for the deeplearnjs website? If so it sounds like @yining1023 might be interested in making that.

**Comment by nsthorat** [comment]
_From @yining1023 on December 30, 2017 22:12_

Yes! I just tried your demo. it's so much fun! it's amazing. It takes about 4 seconds to generate a new image on my computer. I think it's fast enough.

I'm starting making one for the deeplearnjs website. Thank you some much for making this!!
<img width="772" alt="screen shot 2017-12-30 at 17 05 16" src="https://user-images.githubusercontent.com/8662372/34457605-b4353fb8-ed83-11e7-9c38-9e2ed507d069.png">

**Comment by nsthorat** [comment]
We'd definitely be open to linking to the demo from the homepage! We've amassed too many "official" demos (ones that are hosted in our repo), and it's catching up to us now. I'd love if one of you guys could host it and we could link off to you from our homepage and give you total credit - github pages free hosting would work.

I think it would be *awesome* if it were small enough / fast enough in inference that it generates as you are drawing - this would differentiate it from the existing demo (other than of course getting rid of Christopher having to run a TensorFlow instance somewhere). I think it's possible if someone keeps trying to continue on the model you modified (increase depth, use channel multiplier to increase statistical power on separable convs).

If you want to just create a UI around this with the current latency because of lack of time, I understand, but it might be nice in the UI to say something about how we'd love for somebody to take this model and try to make it smaller with some of the things we talked about on this thread.

**Comment by nsthorat** [comment]
_From @yining1023 on January 2, 2018 17:11_

@nsthorat I can host it on my github.
I agree! It would be very cool if it can update as people are drawing. I will try the suggestions from this thread to make the model smaller or faster. I might have some questions along the way. Will come back and ask you guys.

**Comment by nsthorat** [comment]
_From @christopherhesse on January 3, 2018 8:17_

I think you can probably do fine with a few tweaks to the model, but another option might be making convolutions fast.  https://github.com/Maratyszcza/NNPACK for instance seems to support being compiled to webassembly, though I have not looked at it more than that.

**Comment by nsthorat** [comment]
Interesting, not sure how much that will help here though (GPU / CPU transfer overhead is non-negligible, and WASM doesn't take advantage of SIMD yet).

**Comment by nsthorat** [comment]
In general, people are trending towards these type of mobile convolutional kernels. You just need to tweak the hyperparameters. Soon, architecture search will save us :)

**Comment by nsthorat** [comment]
_From @christopherhesse on January 4, 2018 23:47_

Whoops I thought it at least it supported SIMD, I guess it's only 2018.

**Comment by nsthorat** [comment]
I hope it happens. When it does, we're going to introduce a WebAssembly backend.

**Comment by nsthorat** [comment]
This happened in ML5!
